### PR TITLE
feat(feeds): #1750 PageModProvider, add broadcasting/send to feeds

### PR DESCRIPTION
**Pull request by k88hudson**
This patch refactors the worker management to use a `Map` with unique identifiers, so actions from each worker are tagged with a `workerId`. 

Actually using send/broadcast in feeds will be implemented in https://github.com/mozilla/activity-stream/issues/1829

### CODE REVIEW

File: content-test/addon/PageModProvider.test.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+          // This is the callback for a CONTENT_TO_ADDON message
+          return worker.port.on.firstCall.args[1];
+        }
+        it("should remove the worker if the event type is 'pagehide'", () => {
+          const callback = setupWithOnMessage();
+          assert.ok(instance.workers.get(worker));
+          callback({type: "pagehide"});
+          assert.isUndefined(instance.workers.get(worker), "pagehide event removes worker");
+        });
+        it("should remove the worker if the event type is 'pagehide'", () => {
```
Comment (sarracini) [inline content-test/addon/PageModProvider.test.js]: dupe of the test above?

File: content-test/addon/PageModProvider.test.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    });
+  });
+  describe("#getWorkerById", () => {
+    it("should get a worker by id", () => {
+      const id = "asdkl123";
+      const worker = {};
+      instance.workers.set(worker, id);
+      instance.workers.set({}, "wjklwlk");
+      instance.workers.set({}, "a3elaks");
+      assert.equal(instance.getWorkerById(id), worker);
```
Comment (sarracini) [inline content-test/addon/PageModProvider.test.js]: It's not super clear to me how this test is testing that it got the correct worker based on the id... if all the workers are just an empty object, they'll all pass this test no matter what id you pass in? Unless you're just testing that the ```getWorkerById``` function returns any worker?

File: addon/PageModProvider.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   * Adds a worker and returns the new id
+   */
+  addWorker(worker) {
+    if (this.workers.has(worker)) {
+      return this.workers.get(worker);
+    }
+    const id = this.uuid();
+    this.workers.set(worker, id);
+    if (this._logEvent) {
+      this._logEvent(worker.tab, "WORKER_ATTACHED");
```
Comment (sarracini) [inline addon/PageModProvider.js]: Maybe add ```WORKER_ATTACHED``` to constants? That way we don't have to update the tests if it changes

File: content-test/addon/PageModProvider.test.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      instance = new PageModProvider({onAddWorker});
+      instance.addWorker({});
+      assert.calledOnce(onAddWorker);
+    });
+    it("should call logEvent if it was passed to the init function", () => {
+      const logEvent = sinon.spy();
+      const worker = {tab: {}};
+      instance.init({logEvent});
+      instance.addWorker(worker);
+      assert.calledWith(logEvent, worker.tab, "WORKER_ATTACHED");
```
Comment (sarracini) [inline content-test/addon/PageModProvider.test.js]: grab this from constants, once it's added

File: content-test/addon/PageModProvider.test.js (hunk 2 above)
Comment (k88hudson) [inline content-test/addon/PageModProvider.test.js]: `{} !== {}`, so only the reference to the actual object cached with `worker` will actually be correct

Comment (sarracini) [inline content-test/addon/PageModProvider.test.js]: oh javascript, you got me again

**Comment by k88hudson** [comment]
@sarracini ok, changes in https://github.com/mozilla/activity-stream/pull/1830/commits/388c28272eae1ee05aaa1a556a1acb3c390ff3d4

**Pull request merged**